p2p/ipld/read.go

```diff
@@ -0,0 +1,96 @@
+package ipld
+
+import (
+	"context"
+	"errors"
+	"math"
+
+	"github.com/ipfs/go-cid"
+	coreiface "github.com/ipfs/interface-go-ipfs-core"
+	"github.com/ipfs/interface-go-ipfs-core/path"
+)
+
+// /////////////////////////////////////
+//	Get Leaf Data
+// /////////////////////////////////////
+
+// GetLeafData fetches and returns the data for leaf leafIndex of root rootCid.
+// It stops and returns an error if the provided context is cancelled before
+// finishing
+func GetLeafData(
+	ctx context.Context,
+	rootCid cid.Cid,
+	leafIndex uint32,
+	totalLeafs uint32, // this corresponds to the extended square width
+	api coreiface.CoreAPI,
+) ([]byte, error) {
+	// calculate the path to the leaf
+	leafPath, err := calcCIDPath(leafIndex, totalLeafs)
+	if err != nil {
+		return nil, err
+	}
+
+	// use the root cid and the leafPath to create an ipld path
+	p := path.Join(path.IpldPath(rootCid), leafPath...)
+
+	// resolve the path
+	node, err := api.ResolveNode(ctx, p)
+	if err != nil {
+		return nil, err
+	}
+
+	// return the leaf, without the nmt-leaf-or-node byte
+	return node.RawData()[1:], nil
+}
+
+func calcCIDPath(index, total uint32) ([]string, error) {
```

Comment:
Is there any reason this returns a slice of strings instead of a string directly? As far as I understand, these will be joined together anyways. But maybe it is cleaner, as this abstracts away the [separator](https://github.com/ipfs/interface-go-ipfs-core/blob/b935dfe5375eac7ea3c65b14b3f9a0242861d0b3/path/path.go#L118) `/` 🤔

Reply:
yeah, it's just that `[]string` is what [`path.Join`](https://github.com/ipfs/interface-go-ipfs-core/blob/b935dfe5375eac7ea3c65b14b3f9a0242861d0b3/path/path.go#L117) takes. The IPFS path package used here uses some unexported structs to handle paths/resolved nodes and AFAICT this is actually the only way to create that underlying struct in the way that we need. There might be another way using a different package. I could also split a single string before using `path.Join`, but I figured we might as well avoid extra step.

Reply:
It looks like you could also just join the two strings: cid and the already put together path (e.g. 0/1/1/0). But it doesn't matter we can circle back on this later.